add a spec for initial position and launch mode (#2654)
microsoft/terminal: doc/specs/#1043 - Set the initial position of the Terminal/spec.md

Three new properties should be added in the json settings file:

**initialX**: int. This set the initial horizontal position of the top-left corner of the window. This property is optional. If not provided, system default position will be used. 

**initialY**: int. This set the initial vertical position of the top-left corner of the window. This property is optional. If not provided, system default position will be used.

**launchMode**: string. Determine the launch mode. There are two modes for now
                         1. maximize: the window will be maximized when launch.
                         2. default: the window will be initialized with system default size. 

The steps of this process:

1. Set the top-left origin, width and height to CW_USEDEFAULT.
2. Get the dpi of the nearest monitor; Load settings.
3. From settings, find the user-defined initial position and launch mode. 
4. If the user sets custom initial position, calculate the new position considering the current dpi and monitor. If not, use system default value.

> 5. If the user set launch mode as "maximize", calculate the new height and width. If the user choose "default", use system default size.

Reviewer: Will the position be relative to whatever monitor the system selected? The user will not be able to control the monitor in use, correct?

Author: Yes, I think so. In settings->Display, we can see how the monitors are arranged in the system: 

![image](https://user-images.githubusercontent.com/11893272/66610907-d17d6a80-eb71-11e9-9c89-73dbc267969d.png)

And the coordinates depend on this figure. It is also related where the main display is. The top-left corner of the main display is (0, 0). We can rearrange the layout of the monitors or change the main display in settings.

6. SetWindowPos with the new position and dimension of the window. 

> Step 2 to 6 should be done in `AppHost::_HandleCreateWindow`, which is consistent to the current code.

Reviewer: I was wondering if we should simplify the setting for the profiles. Is there ever a reason when you would want an initial position _and_ a maximized window at the same time?

This might also simplify the code for storing the user's profile settings.

What if we did this?

```jsonc
{
    "initialPosition": "100, 50" // combine both numbers into one value
}
```

or

```jsonc
{
    "initialPosition": "maximized" // no numbers, just "maximized".
    // this means you cannot say "I want a 100x100 window that is maximized"
}
```

```jsonc
{
    "initialPosition": null // this is actually the default value, and it means the same as "default"
}
```

Reviewer: Or, is this how the user specifies "Maximize on the left monitor":

-100, 0, maximized?

If we can't combine all three into one value, should we combine the x and y into one string? It may make life easier for the user.

What do you think?

```jsonc
{
    "initialPosition": "100, 50", // combine both numbers into one value
    "launchMode": "maximize"
}
```

Author: I think we also need the initial position in a maximized state and yes, because this is how the user specifies "Maximize on the left monitor": -100, 0, maximized

So I think we can combine the x and y into one string: 

"initialPosition": "100, 50", // combine both numbers into one value
"launchMode": "maximize"

## Abstract

This spec is for task #1043 �Be able to set an initial position for the terminal�. It goes over the details of a new feature that allows users to set the initial position and size of the terminal. Expected behavior and design of this feature is included. Besides, future possible follow-up works are also addressed.

## Inspiration

The idea is to allow users to set the initial position of the Terminal when they launch it, prevent the Terminal from appearing on unexpected position (e.g. outside of the screen bounds). We are also going to let users choose to maximize the window when they launch it.

## Solution Design

For now, the Terminal window is put on a default initial position. The program uses CW_USEDEFAULT in the screen coordinates for top-left corner. We have two different types of window � client window and non-client window. However, code path for window creation (WM_CREATE message is shared by the two types of windows) are almost the same for the two types of windows, except that there are some differences in calculation of the width and height of the window.

Four new properties should be added in the json settings file:

> **useDefaultInitialPos**: true or false. Determine if we use CW_USEDEFAULT as the initial position of the Terminal window. If true, we ignore initialX and initially. Default is true;

Reviewer 2: Do we need this setting? Could the following setting both just be optional settings, and if they exist, assume that for that dimension, we want to use the given position instead of the default?

Author: removed. And I also think if the user only provide one axis value, we use system default for the other one ?

**initialX**: int. If useDefaultInitialPos is false, this set the initial horizontal position of the top-left corner of the window. Default is 0.

> **initiallY**: int. If useDefaultInitialPos is false, this set the initial vertical position of the top-left corner of the window. Default is 0.

Reviewer 2: ```suggestion
**initialY**: int. If useDefaultInitialPos is false, this set the initial vertical position of the top-left corner of the window. Default is 0.
```

Author: removed. Sorry for this mistake

**maximizeWhenLaunch**: true or false. Determine if the window should be maximized upon launch. If true, we ignore initialCols and initialRows. Default is false.

> The flow chart of this process:

Reviewer 2: I'm not sure that having a png for this flow chart really makes the most sense here - there's not really any branching or looping in the flow that would make a flow chart superior to a plain-old numbered list of steps

Author: changed

> ![Sol Design](images/FlowChart.png)

Reviewer 2: Looks like the file is currently just named "Image1.png", that should probably get a better name

Author: removed

> Step 2 to 6 should be done in AppHost::_HandleCreateWindow, which is consistent to the current code.

Reviewer 2: ```suggestion
Step 2 to 6 should be done in `AppHost::_HandleCreateWindow`, which is consistent to the current code.
```

Author: added

In step 4, we may need to consider the dpi of the current monitor and multi-monitor scenario when calculating the initial position of the window.

## UI/UX Design

Upon successful implementation, the user is able to add new properties to the json profile file, which is illustrated in the picture below:

![Sol Design](images/ProfileSnapshot.png)

The rest of the UI will be the same of the current Terminal experience, except that the initial position may be different.

## Capabilities

### Accessibility

This feature will not impact accessibility of Windows Terminal.

### Security

This should not introduce any new security issues.

### Reliability

> This new feathre allows the users to set custom initial position of the Terminal App window, which helps them to avoid embarassing window launch situation such as launching outside the screen bounds. Thus, it improves the reliability.

Reviewer 2: ```suggestion
This new feature allows the users to set custom initial position of the Terminal App window, which helps them to avoid embarrassing window launch situation such as launching outside the screen bounds. Thus, it improves the reliability.
```

Author: changed

### Compatibility

This feature won't break existing features of Terminal.

### Performance, Power, and Efficiency

More data reading and calculation will be included in Terminal Launch process, which may inversely influence the launch time. However, the impact is trivial. 

## Potential Issues

1. The dpi of the monitor may impact the initial position even if the json profile settings are the same. We need to consider the dpi of the user's monitor and calculate the initial position in the current screen coordinates. Testing with different monitor dpi are also necessary.

2. We need to consider multi-monitor scenario. If the user has multiple monitors, we must guarantee that the Terminal could be iniitalized as expected. More discussions on what should we do in this scenario with the github community is needed. 

## Future considerations

For now, this feature only allows the user to set initial positon and choose whether to maximize the window when launch. In the future, we may consider follow-up features like:

> 1. Save the position of the Terminal on exit, and restore the position on the next launch. This could be a true/false feature that users could choose to set.

Reviewer 2: You may want to link this to #766

Author: added

2. We may need to consider multiple Terminal windows scenario. If the user opens multiple Terminal windows, then we need to consider how to save and restore the position. 

> 

Reviewer 2: One thing to also consider probably as a part of this design is supporting launching straight into not only maximized, but also full-screen mode. While we don't support full-screen mode yet, we will by 1.0, and there have already been asks to add support for launching into fullscreen mode by default. 

Having two boolean settings `maximizeWhenLaunched` and `fullscreenWhenLaunched` seems wrong to me - what if they're both true? What if we'd want to add an additional `minimizeWhenLaunched`?

Author: What if we add a settings called "launchMode", and have "default", "maximize" as its property for now. We may also add "minimize", "fullscreen" in the future ? I wonder if these property values are too difficult for users to remember ?

Reviewer 2: Let's summon the wordsmiths and see what input they have @bitcrazed @cinnamon-msft

Author: changed

## Solution Design

For now, the Terminal window is put on a default initial position. The program uses CW_USEDEFAULT in the screen coordinates for top-left corner. We have two different types of window � client window and non-client window. However, code path for window creation (WM_CREATE message is shared by the two types of windows) are almost the same for the two types of windows, except that there are some differences in calculation of the width and height of the window.

Two new properties should be added in the json settings file:

**initialPosition**: string. This sets the initial horizontal and vertical position of the top-left corner of the window. This property follows a structure: "X value, Y value" and has following rules:

1. All spaces will be ignored.

2. Both X value and Y values are optional. If anyone of them is missing, or the value is invalid, system default value will be used. Examples:

       ", 1000" equals to (default, 1000)
       "1000, " equals to (1000, default)
       "," equals to (default, default)
       "abc, 1000" equals to (default, 1000)

**launchMode**: string. Determine the launch mode. There are two modes for now

>                          1. maximize: the window will be maximized when launch.

Reviewer 2: I'd probably add a newline above these two lines, and de-indent them. Right now, in the rendered markdown, they just flow into the above line.

Author: changed
